server/core/hot_region_storage.go

```diff
@@ -0,0 +1,361 @@
+// it will return to tidb
+type HistoryHotRegions struct {
+	HistoryHotRegion []*HistoryHotRegion `json:"history_hot_region"`
+}
+
+// HistoryHotRegion wraps hot region info
+// it is storage format of hot_region_storage
+type HistoryHotRegion struct {
+	UpdateTime    int64   `json:"update_time,omitempty"`
+	RegionID      uint64  `json:"region_id,omitempty"`
+	PeerID        uint64  `json:"peer_id,omitempty"`
+	StoreID       uint64  `json:"store_id,omitempty"`
+	IsLeader      bool    `json:"is_leader,omitempty"`
+	IsLearner     bool    `json:"is_learner,omitempty"`
+	HotRegionType string  `json:"hot_region_type,omitempty"`
+	HotDegree     int64   `json:"hot_degree,omitempty"`
+	FlowBytes     float64 `json:"flow_bytes,omitempty"`
+	KeyRate       float64 `json:"key_rate,omitempty"`
+	QueryRate     float64 `json:"query_rate,omitempty"`
+	StartKey      []byte  `json:"start_key,omitempty"`
+	EndKey        []byte  `json:"end_key,omitempty"`
+	// Encryption metadata for start_key and end_key. encryption_meta.iv is IV for start_key.
+	// IV for end_key is calculated from (encryption_meta.iv + len(start_key)).
+	// The field is only used by PD and should be ignored otherwise.
+	// If encryption_meta is empty (i.e. nil), it means start_key and end_key are unencrypted.
+	EncryptionMeta *encryptionpb.EncryptionMeta `json:"encryption_meta,omitempty"`
+}
+
+type HotRegionStorageHandler interface {
+	// PackHistoryHotWriteRegions get read hot region info in HistoryHotRegion from
+	PackHistoryHotReadRegions() ([]HistoryHotRegion, error)
+	// PackHistoryHotWriteRegions get write hot region info in HistoryHotRegion from
+	PackHistoryHotWriteRegions() ([]HistoryHotRegion, error)
+	IsLeader() bool
+}
+
+const (
+	// leveldb will run compaction after 30 times delete.
+	defaultCompactionTime = 30
+	// delete will run at this o`clock.
+	defaultDeleteTime = 4
+)
+
+// HotRegionTypes stands for hot type.
+var HotRegionTypes = []string{
+	"read",
+	"write",
+}
+
+// NewHotRegionsStorage create storage to store hot regions info.
+func NewHotRegionsStorage(
+	ctx context.Context,
+	path string,
+	encryptionKeyManager *encryptionkm.KeyManager,
+	hotRegionStorageHandler HotRegionStorageHandler,
+	remianedDays int64,
+	pullInterval time.Duration,
+) (*HotRegionStorage, error) {
+	levelDB, err := kv.NewLeveldbKV(path)
+	if err != nil {
+		return nil, err
+	}
+	hotRegionInfoCtx, hotRegionInfoCancle := context.WithCancel(ctx)
+	h := HotRegionStorage{
+		LeveldbKV:               levelDB,
+		encryptionKeyManager:    encryptionKeyManager,
+		batchHotInfo:            make(map[string]*HistoryHotRegion),
+		remianedDays:            remianedDays,
+		pullInterval:            pullInterval,
+		compactionCountdown:     defaultCompactionTime,
+		hotRegionInfoCtx:        hotRegionInfoCtx,
+		hotRegionInfoCancel:     hotRegionInfoCancle,
+		hotRegionStorageHandler: hotRegionStorageHandler,
+	}
+	if remianedDays > 0 {
+		h.hotRegionLoopWg.Add(2)
+		h.backgroundFlush()
+		h.backgroundDelete()
```

Comment:
I'd like to propose to move `go` in methods out to make goroutine management more clear:
```
wg.Add(2)
go h.method1()
go h.method2()
```

Reply:
apply to code